File: src/PHP/BehavioralPatterns/State/VendingMachine/DispenseProductState.php
<?php

namespace BehavioralPatterns\State\VendingMachine;

class DispenseProductState extends State {
    public function __construct(VendingMachine $vendingMachine) {
        parent::__construct($vendingMachine);
        echo "DISPENSING\n";
    }

    public function insertMoney(float $amount): void {
        echo "Cannot insert money during the dispensing of the product\n";
    }

    public function selectProduct(string $productCode): void {
        echo "The product is already selected.\n";
    }

    public function dispenseProduct(): void {
        if ($this->vendingMachine->getSelectedProductCode() === null) {
            echo "There is no selected product for dispensing\n";
            $this->vendingMachine->setState(new IdleState($this->vendingMachine));
            return;
        }

        echo "Dispensing product.\n";
        usleep(2000000); // Sleep for 2 seconds (in microseconds)

        $products = $this->vendingMachine->getProducts();
        $selectedProduct = $products[$this->vendingMachine->getSelectedProductCode()];
        $selectedProduct->setStock($selectedProduct->getStock() - 1);

        $this->vendingMachine->setSelectedProductCode(null);

        echo "The product is dispensed.\n";

        if (array_reduce($products, function ($carry, $product) {
            return $carry && ($product->getStock() == 0);
        }, true)) {
            $this->vendingMachine->setState(new SoldOutState($this->vendingMachine));
        } else {
            $this->vendingMachine->setState(new IdleState($this->vendingMachine));
        }
    }

    public function cancel(): void {
        echo "Cannot cancel the dispensing operation\n";
    }

    public function refill(array $products): void {
        echo "Cannot refill during the dispensing of the product\n";
    }
}

File: src/Java/BehavioralPatterns/State/VendingMachine/src/DispenseProductState.java
import java.util.List;
import java.util.concurrent.TimeUnit;

public class DispenseProductState extends State {
    public DispenseProductState(VendingMachine vendingMachine) {
        super(vendingMachine);
        System.out.println("DISPENSING");
    }

    @Override
    public void insertMoney(double amount) {
        System.out.println("Cannot insert money during the dispensing of the product");
    }

    @Override
    public void selectProduct(String productCode) {
        System.out.println("The product is already selected.");
    }

    @Override
    public void dispenseProduct() {
        if (vendingMachine.getSelectedProductCode() == null) {
            System.out.println("There is no selected product for dispensing");
            vendingMachine.setState(new IdleState(vendingMachine));
            return;
        }

        System.out.println("Dispensing product.");
        try {
            TimeUnit.SECONDS.sleep(2); // Simulating dispensing time
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        Product product = vendingMachine.getProducts().stream()
                .filter(p -> p.getCode().equals(vendingMachine.getSelectedProductCode()))
                .findFirst()
                .orElse(null);

        if (product != null) {
            product.setStock(product.getStock() - 1);
        }

        vendingMachine.setSelectedProductCode(null);

        System.out.println("The product is dispensed.");

        if (vendingMachine.getProducts().stream().allMatch(p -> p.getStock() == 0)) {
            vendingMachine.setState(new SoldOutState(vendingMachine));
        } else {
            vendingMachine.setState(new IdleState(vendingMachine));
        }
    }

    @Override
    public void cancel() {
        System.out.println("Cannot cancel the dispensing operation");
    }

    @Override
    public void refill(List<Product> products) {
        System.out.println("Cannot refill during the dispensing of the product");
    }
}
